gherkin-utils/java/gherkin-utils/src/main/java/gherkin/utils/walker/GherkinDocumentWalker.java

```diff
@@ -0,0 +1,397 @@
+
+        if (feature == null) {
+            return null;
+        }
+
+        return GherkinDocument
+                .newBuilder()
+                .setFeature(feature)
+                .addAllComments(gherkinDocument.getCommentsList())
+                .setUri(gherkinDocument.getUri())
+                .build();
+    }
+
+    private GherkinDocument.Feature walkFeature(GherkinDocument.Feature feature) {
+        List<GherkinDocument.Feature.FeatureChild> keptChildren = walkFeatureChildren(feature.getChildrenList());
+
+        handlers.handleFeature(feature);
+
+        boolean backgroundKept = keptChildren
+                .stream()
+                .anyMatch(featureChild -> featureChild.hasBackground());
+
+        if (filters.acceptFeature(feature) || backgroundKept) {
+            return copyFeature(feature,
+                    feature.getChildrenList()
+                            .stream()
+                            .map(
+                                    child -> {
+                                        if (child.hasBackground()) {
+                                            return GherkinDocument.Feature.FeatureChild
+                                                    .newBuilder()
+                                                    .setBackground(copyBackground(child.getBackground()))
+                                                    .build();
+                                        }
+
+                                        if (child.hasScenario()) {
+                                            return GherkinDocument.Feature.FeatureChild
+                                                    .newBuilder()
+                                                    .setScenario(copyScenario(child.getScenario()))
+                                                    .build();
+                                        }
+
+                                        if (child.hasRule()) {
+                                            return GherkinDocument.Feature.FeatureChild
+                                                    .newBuilder()
+                                                    .setRule(copyRule(child.getRule(), child.getRule().getChildrenList()))
+                                                    .build();
+                                        }
+
+                                        return child;
+                                    }
+                            )
+                            .collect(Collectors.toList())
+            );
+        }
+
+        if (keptChildren
+                .stream()
+                .anyMatch(featureChild -> featureChild != null)) {
+            return copyFeature(feature, keptChildren);
+        }
+
+        return null;
+    }
+
+    private GherkinDocument.Feature copyFeature(GherkinDocument.Feature feature,
+                                                List<GherkinDocument.Feature.FeatureChild> featureChildren) {
+        return GherkinDocument.Feature
```

Comment:
These copy method looks very fragile. When a property is added to gherkin we will not copy it automatically. Consider adding copy-constructors in `messages`.